File: slicing-ios/SlicingGame/Components/NavigationBars/GameTitleBarView.swift
//
//  GameTitleBarView.swift
//  Navigation bar: the title bar for in-game
//

import UIKit
import UniLayout
import JsonInflator

class GameTitleBarView: LinearContainerView, NavigationBarComponent {

    // --
    // MARK: Layout resource
    // --
    
    private let layoutFile = "GameTitleBar"


    // --
    // MARK: Bound views
    // --
    
    private var statusUnderlayView: UniView?
    private var contentView: LinearContainerView?
    private var titleView: TextView?
    private var menuButtonView: ImageButtonView?
    private var actionButtonView: ImageButtonView?
    private var dividerView: UniView?


    // --
    // MARK: Viewlet integration
    // --
    
    override class func viewlet() -> JsonInflatable {
        return ViewletClass()
    }

    private class ViewletClass: JsonInflatable {

        func create() -> Any {
            return GameTitleBarView()
        }

        func update(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any], parent: Any?, binder: InflatorBinder?) -> Bool {
            if let titleBar = object as? GameTitleBarView {
                // Apply title
                titleBar.title = convUtil.asString(value: attributes["localizedTitle"])?.localized() ?? convUtil.asString(value: attributes["title"])
                
                // Apply icons and events
                titleBar.menuIcon = ImageSource.fromValue(value: attributes["menuIcon"])
                titleBar.menuEvents = AppEvent.fromValues(values: attributes["menuEvents"] ?? attributes["menuEvent"])
                titleBar.actionIcon = ImageSource.fromValue(value: attributes["actionIcon"])
                titleBar.actionEvents = AppEvent.fromValues(values: attributes["actionEvents"] ?? attributes["actionEvent"])

                // Apply bar styling
                titleBar.showDivider = convUtil.asBool(value: attributes["showDivider"]) ?? false

                // Generic view properties
                ViewletUtil.applyGenericViewAttributes(convUtil: convUtil, view: titleBar, attributes: attributes)

                // Chain event observer
                if let eventObserver = parent as? AppEventObserver {
                    titleBar.eventObserver = eventObserver
                }
                return true
            }
            return false
        }

        func canRecycle(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any]) -> Bool {
            return type(of: object) == GameTitleBarView.self
        }

    }


    // --
    // MARK: Initialization
    // --
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }

    private func setup() {
        // Inflate
        let binder = InflatorDictBinder()
        ViewletUtil.assertInflateOn(view: self, attributes: JsonLoader.shared.attributesFrom(jsonFile: layoutFile), parent: nil, binder: binder)

        // Bind views
        statusUnderlayView = binder.findByReference("statusUnderlay") as? UniView
        contentView = binder.findByReference("content") as? LinearContainerView
        titleView = binder.findByReference("title") as? TextView
        menuButtonView = binder.findByReference("menuButton") as? ImageButtonView
        actionButtonView = binder.findByReference("actionButton") as? ImageButtonView
        dividerView = binder.findByReference("divider") as? UniView
    }


    // --
    // MARK: Configurable values
    // --
    
    override var backgroundColor: UIColor? {
        set {
            // Apply color to bar
            statusUnderlayView?.backgroundColor = newValue
            contentView?.backgroundColor = newValue
            
            // Colorize title components
            let titleColor = averageColor?.intensity() ?? 0 < 0.25 ? UIColor.white : UIColor.text
            titleView?.textColor = titleColor
            menuButtonView?.colorize = titleColor
            actionButtonView?.colorize = titleColor
            menuButtonView?.highlightedColorize = titleColor.withAlphaComponent(0.5)
            actionButtonView?.highlightedColorize = titleColor.withAlphaComponent(0.5)
            menuButtonView?.disabledColorize = titleColor.withAlphaComponent(0.25)
            actionButtonView?.disabledColorize = titleColor.withAlphaComponent(0.25)
        }
        get { return statusUnderlayView?.backgroundColor }
    }
    
    var title: String? {
        set {
            titleView?.text = newValue
        }
        get { return titleView?.text }
    }
    
    var menuIcon: ImageSource? {
        set {
            let titlePadding: CGFloat = newValue != nil || actionIcon != nil ? 0 : 16
            menuButtonView?.source = newValue
            menuButtonView?.visibility = newValue != nil || actionIcon != nil ? .visible : .hidden
            actionButtonView?.visibility = newValue != nil || actionIcon != nil ? .visible : .hidden
            titleView?.padding = UIEdgeInsets(top: 0, left: titlePadding, bottom: 0, right: titlePadding)
        }
        get { return menuButtonView?.source }
    }
    
    var menuEvents: [AppEvent] {
        set {
            menuButtonView?.tapEvents = newValue
        }
        get { return menuButtonView?.tapEvents ?? [] }
    }

    var actionIcon: ImageSource? {
        set {
            let titlePadding: CGFloat = newValue != nil || actionIcon != nil ? 0 : 16
            actionButtonView?.source = newValue
            menuButtonView?.visibility = newValue != nil || menuIcon != nil ? .visible : .hidden
            actionButtonView?.visibility = newValue != nil || menuIcon != nil ? .visible : .hidden
            titleView?.padding = UIEdgeInsets(top: 0, left: titlePadding, bottom: 0, right: titlePadding)
        }
        get { return actionButtonView?.source }
    }
    
    var actionEvents: [AppEvent] {
        set {
            actionButtonView?.tapEvents = newValue
        }
        get { return actionButtonView?.tapEvents ?? [] }
    }

    var showDivider: Bool = false {
        didSet {
            dividerView?.visibility = showDivider ? .visible : .hidden
        }
    }
    
    
    // --
    // MARK: NavigationBarComponent implementation
    // --
    
    var averageColor: UIColor? {
        get {
            return statusUnderlayView?.backgroundColor
        }
    }

    var statusBarHeight: CGFloat = 0 {
        didSet {
            statusUnderlayView?.layoutProperties.height = statusBarHeight
        }
    }

    var barContentHeight: CGFloat = 0 {
        didSet {
            contentView?.layoutProperties.height = barContentHeight
            menuButtonView?.layoutProperties.width = barContentHeight
            actionButtonView?.layoutProperties.width = barContentHeight
        }
    }

}

File: slicing-android/app/src/main/java/com/crescentflare/slicinggame/components/navigationbars/GameTitleBarView.kt
package com.crescentflare.slicinggame.components.navigationbars

import android.annotation.TargetApi
import android.content.Context
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Build
import android.util.AttributeSet
import android.view.View
import androidx.core.content.ContextCompat
import com.crescentflare.jsoninflator.JsonInflatable
import com.crescentflare.jsoninflator.JsonLoader
import com.crescentflare.jsoninflator.binder.InflatableRef
import com.crescentflare.jsoninflator.binder.InflatorAnnotationBinder
import com.crescentflare.jsoninflator.binder.InflatorBinder
import com.crescentflare.jsoninflator.utility.InflatorMapUtil
import com.crescentflare.slicinggame.R
import com.crescentflare.slicinggame.components.basicviews.TextView
import com.crescentflare.slicinggame.components.containers.LinearContainerView
import com.crescentflare.slicinggame.components.types.NavigationBarComponent
import com.crescentflare.slicinggame.components.utility.ViewletUtil
import com.crescentflare.slicinggame.infrastructure.coreextensions.colorIntensity
import com.crescentflare.slicinggame.infrastructure.coreextensions.localized


/**
 * Navigation bar: the title bar for in-game
 */
open class GameTitleBarView : LinearContainerView, NavigationBarComponent {

    // --
    // Statics
    // --

    companion object {

        // --
        // Static: reference to layout resource
        // --

        private const val layoutResource = R.raw.game_title_bar


        // --
        // Static: viewlet integration
        // --

        val viewlet: JsonInflatable = object : JsonInflatable {
            override fun create(context: Context): Any {
                return GameTitleBarView(context)
            }

            override fun update(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>, parent: Any?, binder: InflatorBinder?): Boolean {
                if (obj is GameTitleBarView) {
                    // Apply title
                    obj.title = mapUtil.optionalString(attributes, "localizedTitle", null)?.localized(obj.context) ?: mapUtil.optionalString(attributes, "title", null)

                    // Apply bar styling
                    obj.showDivider = mapUtil.optionalBoolean(attributes, "showDivider", false)

                    // Generic view properties
                    ViewletUtil.applyGenericViewAttributes(mapUtil, obj, attributes)
                    return true
                }
                return false
            }

            override fun canRecycle(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>): Boolean {
                return obj::class == GameTitleBarView::class
            }
        }
    }


    // --
    // Bound views
    // --

    @InflatableRef("statusUnderlay")
    private var statusUnderlayView: View? = null

    @InflatableRef("content")
    private var contentView: View? = null

    @InflatableRef("title")
    private var titleView: TextView? = null

    @InflatableRef("divider")
    private var dividerView: View? = null


    // --
    // Initialization
    // --

    @JvmOverloads
    constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0)
            : super(context, attrs, defStyleAttr)

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int)
            : super(context, attrs, defStyleAttr, defStyleRes)

    init {
        ViewletUtil.assertInflateOn(this, JsonLoader.instance.loadAttributes(context, layoutResource),null, InflatorAnnotationBinder(this))
    }


    // --
    // Configurable values
    // --

    override fun setBackgroundColor(color: Int) {
        statusUnderlayView?.setBackgroundColor(color)
        contentView?.setBackgroundColor(color)
        titleView?.setTextColor(if (averageColor?.colorIntensity() ?: 0.0 < 0.25) Color.WHITE else ContextCompat.getColor(context, R.color.text))
    }

    var title: String?
        get() = titleView?.text?.toString()
        set(title) {
            titleView?.text = title
        }

    var showDivider: Boolean = false
        set(showDivider) {
            field = showDivider
            dividerView?.visibility = if (showDivider) VISIBLE else GONE
        }


    // --
    // NavigationBarComponent implementation
    // --

    override val averageColor: Int?
        get() = (contentView?.background as? ColorDrawable)?.color

    override var statusBarHeight: Int = 0
        set(statusBarHeight) {
            field = statusBarHeight
            statusUnderlayView?.layoutParams?.height = statusBarHeight
        }

    override var barContentHeight: Int = 0
        set(barContentHeight) {
            field = barContentHeight
            contentView?.layoutParams?.height = barContentHeight
        }

}
